### ISSUE TITLE
Provide support for external request signing (extract private key from client)

### ISSUE THREAD

**Issue by mialbu** [nspcc-dev/neofs-sdk-go#373]
Currently, it seems in order to make any request to NeoFS (using the SDK), you have to pass private key data to the client to sign messages. However, I'd like to keep away my private key data from this code and create the necessary signature elsewhere, e.g., in a cold wallet.

For example, to create a new container, I image a process like the following:

1. Prepare a request to create a container and get the (stable marshalled) request data of it.
2. Sign the request data in a separate environment.
3. Build the request based on the request data and the signature data and then send it within the SDK.

Can you tell me if this is somehow possible to do already and I missed it? Otherwise, I think this would be a very useful feature.

@gsmachado @cthulhu-rider

**Comment by gsmachado** [comment]
This is an important "feature", IMHO. 👍

**Comment by roman-khimov** [comment]
This should be supported, I agree, but it may take us some time to get there. It's relatively easy to add some dump/sign/send functionality for NeoFS messages, but I'd like to avoid it.

In general, any NeoFS message is signed and NeoFS core communication model is defined via gRPC. On one hand it's good, clients can use a single communication channel and a unified API for all of their needs, but on the other many of NeoFS APIs just wrap sidechain calls, container creation is a perfect example here --- it's just a transaction on the NeoFS sidechain. And we have _a lot_ of tooling to work with transactions in NeoGo, including [offline signatures](https://github.com/nspcc-dev/neo-go/blob/4e6ce9c3e666dd3979f31287dfdffc7a773544c7/docs/cli.md#offline-signing). 

But it could be used only if the SDK/client is to use Neo RPC, which is a whole different world from the gRPC. In some ways, it's a better world (otherwise there wouldn't be things like nspcc-dev/neofs-net-monitor#74). And in some ways, this world is somewhat incomplete: BLOBs are not just nspcc-dev/neofs-contract#298 thing (they're pretty common in NeoFS contracts), `container.Put()` call needs to be signed by Alphabet nodes (which is related to BLOB problem). If we're to solve these problems, you'd get this feature (and many other ones) for free.

That said, some operations are gRPC-only, for example, object puts/gets, but I doubt you're interested in doing offline signatures for them, it's just not practical (you may want to use some key that is approved for these operations by EACL/Bearer token).

**Comment by cthulhu-rider** [comment]
If I understood correctly you are interested rather in choosing a signing method than in message serialization (which IMO is still better left on the library side). If so, then https://pkg.go.dev/github.com/nspcc-dev/neofs-sdk-go@v1.0.0-rc.7/crypto#Signer has been added to customize the signature calculation mechanism. In the future, it is planned to support it in the client library. I think this will cover the described wishes, correct me if I'm wrong.

**Comment by mialbu** [comment]
> If I understood correctly you are interested rather in choosing a signing method than in message serialization (which IMO is still better left on the library side). If so, then https://pkg.go.dev/github.com/nspcc-dev/neofs-sdk-go@v1.0.0-rc.7/crypto#Signer has been added to customize the signature calculation mechanism. In the future, it is planned to support it in the client library. I think this will cover the described wishes, correct me if I'm wrong.

Hmm, okay, it seems the sign methods are public. However, what I would like to have is a method that let's me create the signature outside of `neofs-sdk-go`. Let me give an example:

For `ContainerPut` ([here](https://github.com/nspcc-dev/neofs-sdk-go/blob/ae44191e8c04a2b527efef6387c128b3b7b33381/client/container.go#L85)) I would like to have another method that does not contain the [signing sequence](https://github.com/nspcc-dev/neofs-sdk-go/blob/ae44191e8c04a2b527efef6387c128b3b7b33381/client/container.go#L99-L108), i.e., does not require a private key. In the [shared library](https://github.com/axlabs/neofs-shared-lib), I'm currently using this method as is. I could rewrite the whole method myself, but afaik I'd run into some problems, since some methods are not public (e.g., [`writeXHeadersToMeta`](https://github.com/nspcc-dev/neofs-sdk-go/blob/ae44191e8c04a2b527efef6387c128b3b7b33381/client/container.go#L117)).

On top of my head two approaches come to mind that could solve the problem:

1) Add another parameter to the client's methods:
```go
func (c *Client) ContainerPut(ctx context.Context, prm PrmContainerPut, sig neofscrypto.Signature) (*ResContainerPut, error) {
```

2) Add another field in the `Prm...` types. For example:

`sig neofscrypto.Signature` as a field in the `PrmContainerPut`

For example, for the latter approach, in the [corresponding code](https://github.com/AxLabs/neofs-shared-lib/blob/f87a9f637d07821d234ed3f52e1c16dc4a7df953/container/container.go#L30-L49) of the shared library we're currently building, we could set the signature as follows:

```go
var prmContainerPut neofsclient.PrmContainerPut
prmContainerPut.SetContainer(*cnr)
pmrContainerPut.SetSignature(&sig)
```

With such a solution, the shared library could provide an approach that let's the developer create the signature outside of the client from `neofs-sdk-go`, i.e., outside of the shared library that we're building. The following sequence would become possible that way:

1. Specify a container with policy, owner, etc. (using e.g., neofs-sdk-go, neow3j, ...) 
2. Get the container's bytes that need to be signed (using e.g., neofs-sdk-go, neow3j, ...)
3. Sign the container bytes (using any wallet you'd like, e.g., neow3j, neo-cli, ...)
4. Use the client from the shared library to execute the `ContainerPut` request providing the container and the required signature

The private key would never have to be consumed by the shared library. Thus, the developer could decide whether to use a wallet application, neo-cli, neow3j, neo-go, or any code to sign the request.

**Comment by cthulhu-rider** [comment]
@mialbu thank u for the explanation, I think I got your point now. Initially, I thought that you needed to custom-sign the requests (I mean not the container signature as part of the `ContainerService.Put` request payload, but the system signatures in the request headers).

In current package API you can't provide pre-calculated container signature, but we can extend `PrmContainerPut` with the option you described in the comment.

**Comment by gsmachado** [comment]
@roman-khimov I believe that the implementation of this issue (through PR https://github.com/nspcc-dev/neofs-sdk-go/pull/398) is not what we asked for...

Instead of setting a `Signer` implementation externally, we just want to get the raw request, sign externally, and then set the signed data back to the SDK.

Am I missing something or this is not possible at all at the current implementation?

**Comment by roman-khimov** [comment]
Try instantiating a https://pkg.go.dev/github.com/nspcc-dev/neofs-sdk-go@v1.0.0-rc.10/crypto#StaticSigner with precalculated signature and let SDK take signature from it.

**Comment by gsmachado** [comment]
@roman-khimov I will check with more time later.

Just another question: How would we have an async signature with the `Signer` interface? Let's imagine that there's a time gap between the signature request and the signature response?

**Comment by mialbu** [comment]
@roman-khimov Basically, we'd like to not just use a different implementation of `Sign()`, but we'd like to create the signature async - this could be in a different process or even on another machine.

So, specifically, we'd like to provide something like following process to the users of neow3j when for example creating a container on NeoFS:

1. Configure the container with `neofs-api-java`.
2. Get the stable marshalled byte array of the container - through the `neofs-shared-lib` indirectly using the `Marshal()` function of `neofs-sdk-go`:

https://github.com/nspcc-dev/neofs-sdk-go/blob/d493dd3e0f51b9dd2bb675c17639198f28b9c87d/container/container.go#L154-L156

3. Sign this byte array wherever desired, e.g., on another machine.
4. Use a method providing the container and the signature to process the call.

What I know is that steps 1-3 are already possible with rc10. However, I'm not sure about the 4th step. What we'd like is to have the client's `ContainerPut` method with a signature parameter instead of the signer interface, i.e., extracting the singing part of the following function, and have an exposed function without the singing -> *.

https://github.com/nspcc-dev/neofs-sdk-go/blob/d493dd3e0f51b9dd2bb675c17639198f28b9c87d/client/container.go#L72-L90

* Basically replacing the above code, with this :point_down:

```golang
func (c *Client) ContainerPut(ctx context.Context, cont container.Container, sig neofscrypto.Signature, prm PrmContainerPut) (cid.ID, error) {
        var cnr v2container.Container 
 	cont.WriteToV2(&cnr)
```

**Comment by mialbu** [comment]
@roman-khimov Basically, we'd like to not just use a different implementation of `Sign()`, but we'd like to create the signature async - this could be in a different process or even on another machine.

So, specifically, we'd like to provide something like following process to the users of neow3j when for example creating a container on NeoFS:

1. Configure the container with `neofs-api-java`.
2. Get the stable marshalled byte array of the container - through the `neofs-shared-lib` indirectly using the `Marshal()` function of `neofs-sdk-go`:

https://github.com/nspcc-dev/neofs-sdk-go/blob/d493dd3e0f51b9dd2bb675c17639198f28b9c87d/container/container.go#L154-L156

3. Sign this byte array wherever desired, e.g., on another machine.
4. Use a method providing the container and the signature to process the call.

What I know is that steps 1-3 are already possible with rc10. However, I'm not sure about the 4th step. What we'd like is to have the client's `ContainerPut` method with a signature parameter instead of the signer interface, i.e., extracting the singing part of the following function, and have an exposed function without the singing -> *.

https://github.com/nspcc-dev/neofs-sdk-go/blob/d493dd3e0f51b9dd2bb675c17639198f28b9c87d/client/container.go#L72-L90

* Basically replacing the above code, with this :point_down:

```golang
func (c *Client) ContainerPut(ctx context.Context, cont container.Container, sig neofscrypto.Signature, prm PrmContainerPut) (cid.ID, error) {
        var cnr v2container.Container 
 	cont.WriteToV2(&cnr)
```

**Comment by roman-khimov** [comment]
In fact steps 1-3 are more problematic than 4. In some cases they're easy (https://pkg.go.dev/github.com/nspcc-dev/neofs-sdk-go@v1.0.0-rc.10/session#Object.SignedData), in others not so much (but #510).

Regarding 4 it's as easy as `NewStaticSigner(scheme, signature, publicKey)`, you do not need private key for it.

**Comment by mialbu** [comment]
Oh ofc, now I understand.
